projects/ad9361/src/main.c

```diff
@@ -715,21 +817,131 @@ int main(void)
+	memset((uint32_t *)address_mem_read,0x00,to_read*2);
+	Xil_DCacheFlushRange((uintptr_t)address_mem_read, to_read*2);
+
+	struct axi_dma_transfer read_transfer = {
+		// Number of bytes to write/read
+		to_read,
+		// Transfer done flag
+		0,
+		// Signal transfer mode
+		NO,
+		// Address of data source
+		0,
+		// Address of data destination
+		(uintptr_t)address_mem_read
+	};
+
+	/* Clear the cache. */
+	Xil_DCacheInvalidateRange((uintptr_t)address_mem_read, to_read);
+
+	/* Read the data from the ADC DMA. */
+	axi_dmac_transfer_start(rx_dmac, &read_transfer);
+
+	/* Wait until transfer finishes */
+	status = axi_dmac_transfer_wait_completion(rx_dmac, 500);
+	if(status < 0)
+		return status;
+
+	/* Clear the cache. */
+	Xil_DCacheInvalidateRange((uintptr_t)address_mem_read, to_read);
+
+	/* Stop DMA rx component */
+	axi_dmac_transfer_stop(rx_dmac);
+	/* Stop DMA tx component */
+	axi_dmac_transfer_stop(tx_dmac);
+
 #endif
 #ifdef XILINX_PLATFORM
 #ifdef FMCOMMS5
 	Xil_DCacheInvalidateRange(ADC_DDR_BASEADDR,
 				  samples * AD9361_ADC_DAC_BYTES_PER_SAMPLE * (ad9361_phy_b->rx_adc->num_channels
 						  +
 						  ad9361_phy->rx_adc->num_channels));
-#else
-	Xil_DCacheInvalidateRange((uintptr_t)adc_buffer, sizeof(adc_buffer));
 #endif
-	printf("DAC_DMA_EXAMPLE: address=%#lx samples=%lu channels=%u bits=%lu\n",
-	       (uintptr_t)adc_buffer, NO_OS_ARRAY_SIZE(adc_buffer), rx_adc_init.num_channels,
-	       8 * sizeof(adc_buffer[0]));
+	printf("DAC_DMA_EXAMPLE: address=%#x samples=%lu channels=%u bits=%u. No. of bytes read from mem: %lu \n",
+	       (uintptr_t)address_mem_read, to_read/8, rx_adc_init.num_channels,
+	       8 * sizeof(uint32_t), to_read/2);
+	/*------------------------------------------------------------------------*/
+#endif
+
+#ifdef XILINX_PLATFORM
+	/* MEM2MEM transfer ------------------------------------------------------*/
+	/* Uncomment if big transfer functionality testing is desired */
+	mem_dmac->max_length = 0x3fff;
+
+	/* Memory address to write constant data to be transmitted */
+	uint32_t src_address_mem = 0x4000000;
+
+	/* Memory address to store read data */
+	uint32_t dest_address_mem = 0xf000000;
+
+	/* Amount of data to read */
+	uint32_t to_transfer_mem = sizeof(uint32_t)*8192*multiplications;
+
+	/* Set contents of memory where we will save the read data to
+	 * all zeros. */
+	memset((uint32_t *)dest_address_mem,0x00,to_transfer_mem);
+	Xil_DCacheFlushRange((uintptr_t)dest_address_mem, to_transfer_mem);
+
+	struct axi_dma_transfer mem_transfer = {
+		// Number of bytes to write/read
+		to_transfer_mem,
+		// Transfer done flag
+		0,
+		// Signal transfer mode
+		NO,
+		// Address of data source
+		(uintptr_t)src_address_mem,
+		// Address of data destination
+		(uintptr_t)dest_address_mem,
+		// Address of data destination
+	};
+
+	/* Clear the cache. */
+	Xil_DCacheInvalidateRange((uintptr_t)dest_address_mem, to_transfer_mem);
+
+	/* Transfer the data. */
+	axi_dmac_transfer_start(mem_dmac, &mem_transfer);
```

Comment:
is always `axi_dmac_transfer_start` followed directly by `axi_dmac_transfer_wait_completion`? If so, I'd suggest squashing these 2 functions into a single one.

Reply:
There is a use case where the DMA transfer is started, some activities are performed, and only after these activities finish executing the completion of the DMA transfer is checked.